### ISSUE TITLE
Add "numerical toggle" selector option with option for debouncing service call

### ISSUE THREAD

**Issue by irakhlin** [Nerwyn/service-call-tile-feature#21]
**Is your feature request related to a problem? Please describe.**
I am trying to replace all tile card built-in features using the service-call option for consistency and customization. This feature would allow things like setting temperature/humidity on a climate entity. 
example of stock feature 'Target temperature':
![image](https://github.com/Nerwyn/service-call-tile-feature/assets/12637936/6fbc7b6e-489c-43e3-9961-c5693828aea4)

**Describe the solution you'd like**
I can see one option to be allowing to set the following configurations:
service_call for toggle up, service_call for toggle down, entity_id and optional attribute for state, a step size, and a min and max value and lastly either a debounce interval or simply a debounce on/off for the service call.

**Describe alternatives you've considered**
I've looked at the following code from home assistant frontend:
https://github.com/home-assistant/frontend/blob/dev/src/panels/lovelace/card-features/hui-target-temperature-card-feature.ts
https://github.com/home-assistant/frontend/blob/dev/src/common/util/debounce.ts

unfortunately without the debounce option this would be very unusable even if trying to style the existing selector option 

**Additional context**
The reason for the debounce being necessary is that many climate entities take a little while to respond to a single service call, so if attempting to change the temperature from 65 to 70, it would be much better to debounce the service calls and only send the final value of 70 as sending a service call for each number results in significant lag.

If this is a feature you will consider adding let me know how I can help.

**Comment by Nerwyn** [comment]
Temporary hacky solution:

```yaml
type: custom:service-call
entries:
  - type: button
    icon: mdi:minus
    label: ' '
    style:
      flex-direction: row
    icon_style:
      padding: 8px
    tap_action:
      service: climate.set_temperature
      data:
        entity_id: climate.downstairs_thermostat
        temperature: '{{ state_attr("climate.downstairs_thermostat", "temperature") - 1 }}'
  - type: button
    label: '{{ state_attr("climate.downstairs_thermostat", "temperature") }} {{ UNIT }}'
    unit_of_measurement: ' °F'
    style:
      flex-basis: 600%
  - type: button
    icon: mdi:plus
    label: ' '
    style:
      flex-direction: row-reverse
    icon_style:
      padding: 8px
    tap_action:
      service: climate.set_temperature
      data:
        entity_id: climate.downstairs_thermostat
        temperature: '{{ state_attr("climate.downstairs_thermostat", "temperature") + 1 }}'
```
![image](https://github.com/Nerwyn/service-call-tile-feature/assets/43071698/514d388c-8f8a-4a0d-bd79-c4826d9e9c95)

**Comment by Nerwyn** [comment]
I've been thinking of adding this (and the dropdown) for a while but hadn't thought through the details like you have. Thanks for that. Here's a snippet of the number buttons HTML

![image](https://github.com/Nerwyn/service-call-tile-feature/assets/43071698/6f49ef24-e48c-4092-9bdd-79dd11bd1251)

It looks like it's two icon buttons with an input range in the middle. I think I can repurpose my existing service call buttons, background, and icon/label elements to create this.

**Comment by Nerwyn** [comment]
The whole debounce and lack of climate increment/decrement does throw a wrench in this, since this tile feature doesn't simply call two different services but instead calls one service and tracks it's internal state, only calling the service once it's internal state stops changing via debounce. it makes me think that I may have to create two different features, one that debounces using a `set_value` like service call and one that calls two different services.

**Comment by Nerwyn** [comment]
Implementation thoughts for debounce version of this feature (since no debounce two service version is easy):
- What advantage does debounce have over just setting a timeout and resetting that timeout every time the value is changed?
- Buttons should:
  - Clear the debounce timeout if defined and set it to undefined.
  - Set [`getValueFromHass`](https://github.com/Nerwyn/service-call-tile-feature/blob/main/src/classes/base-service-call-feature.ts#L27) to false.
  - Decrement/increment internal value.
  - Set a new debounce timeout to call the tap action.
- On timeout complete:
  - Call the set value service call using internal `VALUE`.
  - Set debounce timeout to undefined.
  - Set a timeout to set `getValueFromHass` to true to prevent value bouncing before hass object updates [like slider does](https://github.com/Nerwyn/service-call-tile-feature/blob/main/src/classes/service-call-slider.ts#L139-L142).

**Comment by irakhlin** [comment]
@Nerwyn Thank you for looking into this so deeply, first off the idea of this potentially being two different "implementations" does make sense, many services will have no use for a debounce/timeout especially if two different services are called (one for incrementing and one for decreasing). 

As far as the debounce/timeout solution I also agree that yes, the service will need to be called with the value of the internally maintained VALUE state. My only confusion is the distinction between it using a "timeout" or "debounce". My understanding is that a debounce would be maintained by setting a timeout value and clearing it on each call to the function?

I imagine the functionality would be something like this:
- Currently my climate entity set temperature is: 70,
- I hit the '-' button 5 times, each press is within 1 second of each other
- Each time the '-' is hit the VALUE is decreased by 1, so at this point the VALUE is 65
- When the '-' has not been pressed for 1 second since the last press the call-service is called with the final "VALUE"

Looking at some of the implementations of [debounce](http://underscorejs.org/#debounce) it would prevent the final function (call-service) from being called until user-input has stopped, making sure the function is only called once. If i understand the alternative of using only a simple timeout, it could mean that if a user is pressing the button 5 times and it takes 2 seconds but the timeout is set to 1 the result would potentially be 2 service-calls?

**Comment by Nerwyn** [comment]
Your interpretation of the functionality is correct. I just don't understand why a separate debounce function is necessary when we can instead set, cancel, and reset a timeout. Every time the button is pressed it will first cancel the timeout, then increment/decrement `VALUE`, then set the timeout again. So if the user presses the button five times within two seconds, it will cancel the timeout and restart the timer every time, but only finally call the service one second after the final press.

pseudocode:
```typescript
@customElement('service-call-number-target')
export class ServiceCallNumberTarget extends BaseServiceCallFeature {
    debounceTimer?: ReturnType<typeof setTimeout>
    this.increment: number = 1

    onClick(e: TouchEvent | MouseEvent) {
        this.getValueFromHass = false;
        if (this.debounceTimer) {
            clearTimeout(this.debounceTimer);
            this.debounceTimer = undefined;
        }

        this.value += this.increment
        
        this.debounceTimer = setTimeout(() => {
            this.debounceTimer = undefined;
            this.sendAction('tap_action');
            this.getValueFromHassTimer = setTimeout(() => { 
                this.getValueFromHass = true
            }, 1000);
        }, 1000);
    }
}
```

**Comment by Nerwyn** [comment]
[Here's a jsfiddle of it in action](https://jsfiddle.net/nerwyn/1tdw5vyr/31/). Seems to work fine as described above without the debounce function.

**Comment by irakhlin** [comment]
@Nerwyn , sorry i think i misunderstood your original point about a separate function for the debounce vs the example above. This makes sense, thank you. Do you think it would make sense to have the debounce/timeout configuration value to either take a boolean or number, so if enabled set the default to 1s but also allow setting to a specific ms value? Not a huge deal either way but seems like its not much extra work.

**Comment by Nerwyn** [comment]
It does make sense. That reminds me, I forgot to port over the double tap and hold time customization options from my other project. I should remember to do that in this release.

**Comment by irakhlin** [comment]
> It does make sense. That reminds me, I forgot to port over the double tap and hold time customization options from my other project. I should remember to do that in this release.

Awesome idea, while on the subject of input, I see you are using 'custom-card-helpers' for forwardHaptic but only for the confirmation option, could be a nice (unrelated to this story) feature to allow setting a haptic configuration for standard button presses similar to how custom:button-card does.

**Comment by Nerwyn** [comment]
I hadn't added haptics since the vanilla tile features don't have them, but it wouldn't hurt to add them and also port over that code from android-tv-card.

**Comment by Nerwyn** [comment]
[Here's an even better jsfiddle of it](https://jsfiddle.net/nerwyn/1tdw5vyr/120/). This should be a good starting point for implementing it as a tile feature.

**Comment by Nerwyn** [comment]
The latest alpha has a spinner (name isn't set in stone) that looks and performs like the climate set temperature default tile feature.

```yaml
type: custom:service-call
entries:
  - type: spinner # spinbox is probably a better name
    icon: mdi:thermometer
    label: VALUE
    step: 1 # defaults to 1
    debounce_time: 1000 # defaults to 1000 (ms)
    value_attribute: temperature
    unit_of_measurement: ' °F'
    tap_action:
      service: climate.set_temperature
      data:
        entity_id: climate.downstairs_thermostat
        temperature: VALUE
```

Spinner/Spinbox/ValueSetter/Whatever the final name is TODOs:

- [ ] Make increment and decrement icons customizable.
- [ ] Add optional labels to increment and decrement buttons.
- [ ] Add option for separate service calls for increment and decrement buttons.
- [ ] Make buttons optionally repeat when held.
- [ ] Add swipe left/right functionality.
  - Will have to add exception to swipe navigation project
- [ ] Add configurable max and min values.

Other TODOs:
- [ ] Make hold time, hold repeat interval, and double tap window customizable.
- [ ] Add optional haptics.

**Comment by irakhlin** [comment]
Wow! Amazing stuff, so I setup a test to compare this with to the default climate set temperature feature

![image](https://github.com/Nerwyn/service-call-tile-feature/assets/12637936/5485ad0c-fcd7-4b2a-bf75-9b60998c21e1)

The 4th row down (first 70 F) is the stock implementation of the tile card
The 7th row down (last one) is the new feature **one note, in the example above you have type of spinbox, with this the arrows do not render but type of spinner works** 

The debounce seems to work, clicking + or - multiple times only sends the final value for the call service, the only thing I am noticing is that potentially because of the slight lag of my climate entities (esphome mini splits) to update the value I see the label flash back to the original value, and then quickly to the correct value. So it looks like this:
If the temperature is 67 and I click a few times to 72, it stays at 72, then quickly flashes back to 67 and again to 72, but it seems the call-service is only called once. I don't see this same behavior with the stock/default version. I tried recording a small video of this if sorry about the quality.. seems it scaled weird.



https://github.com/Nerwyn/service-call-tile-feature/assets/12637936/18076802-16bb-4bda-a5c2-908318c855e4

**Comment by Nerwyn** [comment]
> in the example above you have type of spinbox, with this the arrows do not render but type of spinner works
I renamed it to spinbox (since spinner could be confused for something that spins) in the latest alpha, try redownloading it.

> I see the label flash back to the original value, and then quickly to the correct value.
I'll have to increase the `getValueFromHassTimer` timeout length. 1 second seems to be fine for sliders, but not for the spinbox.

I'll keep working on this later tonight.

**Comment by Nerwyn** [comment]
As of alpha 15, the increment and decrement buttons can be stylized and changed to different actions using the Home Assistant actions syntax. I've also added a range option similar to the slider to clamp the min and max values when in debounce mode so you shouldn't run into errors if you go out of range. Here's an example config. The increment and decrement tap actions are set to none below so you can easily switch between debounce and regular action mode by commenting and uncommenting the lines above/below them.

```yaml
type: custom:service-call
entries:
  - type: spinbox
    icon: mdi:brightness-4
    label: VALUE
    step: 10
    debounceTime: 1000
    range:
      - 0
      - 100
    value_attribute: brightness
    tap_action:
      action: call-service
      service: light.turn_on
      data:
        entity_id: light.sunroom_ceiling
        brightness_pct: VALUE
    decrement:
      icon: mdi:brightness-3
      label: down
      tap_action:
        xaction: call-service
        action: none
        service: light.turn_on
        data:
          entity_id: light.sunroom_ceiling
          brightness_step_pct: -10
      style:
        flex-flow: row
      icon_style:
        padding-right: 4px
    increment:
      icon: mdi:brightness-2
      label: up
      tap_action:
        xaction: call-service
        action: none
        service: light.turn_on
        data:
          entity_id: light.sunroom_ceiling
          brightness_step_pct: 10
      style:
        flex-flow: row-reverse
      icon_style:
        padding-left: 4px
    style:
      '--light-color': rgb({{ state_attr("light.sunroom_ceiling", "rgb_color") }})
      '--on-color': >-
        {{ "var(--light-color)" if is_state("light.sunroom_ceiling", "on") else
        "initial" }}
      '--background': var(--on-color)
      '--icon-color': var(--on-color)
      '--label-color': var(--on-color)
```

**Comment by irakhlin** [comment]
This is awesome stuff! 
I am no longer seeing the value bounce issue with the climate entities. Just to share, this is what a sample climate entity now looks like in the on state using only service-call:
![image](https://github.com/Nerwyn/service-call-tile-feature/assets/12637936/19e7092a-a79f-4ce9-94d9-d0ba43b01b95)


off state:
<img width="303" alt="Screenshot 2024-03-27 123958" src="https://github.com/Nerwyn/service-call-tile-feature/assets/12637936/07376e97-0f7b-4209-b843-e091abd69347">

**Comment by Nerwyn** [comment]
I've ported over a bunch of button features from android-tv-card so these features now support support momentary buttons, customizable timings for different actions, repeat on hold, and optional haptics. I've also made it so that if actions are defined specifically for the increment or decrement buttons in the spinbox, it'll render a `service-call-button` instead of a debounced button, so you'll have access to all the features a button feature has.

TODOs:
- [ ] Update README.
- [ ] Add repeat on hold functionality to debounced buttons.
- [ ] Add swipe functionality to spinbox.
- [ ] Slider tooltip style options.
- [ ] [Slider tooltip display scale.](https://github.com/Nerwyn/android-tv-card/issues/74)
- [ ] Unit of measurement option.
